utilities/scripts/python/vendor_files_check.py

```diff
@@ -1,11 +1,188 @@
+            )
+            out = __str_diff__(diff, vendor_file_path, provided_file_path)
+            rc = 1 if diff else 0
+
+            return exec_manager.ProcedureResult(
+                rc,
+                command_name,
+                out,
+            )
+
+        return exec_manager.procedure_run(
+            procedure,
+            command_name,
+            log=log,
+        )
+
+
+@dataclass
+class DiffEntry:
+    """
+    A data class representing an entry in a difference collection.
+    Attributes:
+    val: any
+        The value of the entry.
+    add_or_remove_flag: bool
+        A flag indicating whether the entry should be removed (False) or added (True).
+    """
+
+    val: any
+    add_or_remove_flag: bool
+
+
+def __inclusion_diff__(expected, provided):
+    """
+    Calculate an inclusion diff between the expected and provided inputs in a recursive manner.
+
+    Args:
+        expected: The expected input.
+        provided: The provided input.
+
+    Returns:
+        A dictionary representing the difference between the expected and provided inputs.
+    """
+    diff = {}
+    if not isinstance(expected, dict):
+        if expected != provided:
+            return [DiffEntry(expected, True), DiffEntry(provided, False)]
+    else:
+        for key in expected:
+            if key not in provided:
+                diff[key] = DiffEntry(expected[key], True)
+            else:
+                res = __inclusion_diff__(expected[key], provided[key])
+                if res != {}:
+                    diff[key] = res
+    return diff
+
+
+def __strict_diff__(expected, provided):
+    """
+    Calculate the strict diff between  the expected and provided inputs.
+
+    Args:
+        expected: The expected input for the comparison.
+        provided: The actual input for the comparison.
+
+    Returns:
+        dict: A dictionary representing the strict difference between the expected
+        and provided inputs.
+    """
+
+    def change_flags(diff):
+        if isinstance(diff, DiffEntry):
+            diff.add_or_remove_flag = not diff.add_or_remove_flag
+        if isinstance(diff, list):
+            for val in diff:
+                change_flags(val)
+        if isinstance(diff, dict):
+            for key in diff:
+                change_flags(diff[key])
+
+    # Finds two inclusion diffs and concatenate the results
+    # Also it is important to update a result from the second inclusion diff result
+    # Because it's result has a "reverse" add_or_remove_flag meaning
+    incl1 = __inclusion_diff__(expected, provided)
+    incl2 = __inclusion_diff__(provided, expected)
+    change_flags(incl2)
+    incl1.update(incl2)
+    return incl1
+
+
+def __add_color__(val: str, color: str):
+    if color == "red":
+        return f"\033[91m{val}\033[0m"
+    if color == "green":
+        return f"\033[92m{val}\033[0m"
+    if color == "yellow":
+        return f"\033[93m{val}\033[0m"
+    return val
+
+
+def __str_diff__(
```

Comment:
It would be better if the diff result type was a dataclass, and this was a method of it.
That way, you could easily print it.
The `__inclusion_diff__` and `__strict_diff__` functions would also be better as methods of the dataclass.

For example:

`diff = Diff(expected, provided, strict)`

Internally it handles the `strict` logic that's currently in `toml_diff_check`,
When its created, it can automatically run the diff.
You can then use methods on it to check the result of the diff, or print it out.

Doing this will improve reusability

Reply:
Yea, make sense, updating it

Reply:
updated